File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int max = Integer.MIN_VALUE;

    public int diameterOfBinaryTree(TreeNode root) {
        int res = diameter(root);
        return max;

    }

    public int diameter(TreeNode root) {
        if(root == null) {
            return -1;
        }

        int left =  1 + diameter(root.left);
        int right = 1 + diameter(root.right);

        max = Math.max(left + right, max);

        return Math.max(left, right);
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        return goodNodes(root, -10000);
    }

    int goodNodes(TreeNode* root, int max) {
        if(root == nullptr) {
            return 0;
        }

        int result = (root->val >= max)? 1 : 0;
        
        if(root->val > max) {
            max = root->val;
        }

        int left = goodNodes(root->left, max);
        int right = goodNodes(root->right, max);

        return result + left + right;
    }
};
